### PR TITLE
feat: Validate custom partition templates on their creation

### LINKED ISSUE #7895: Validate custom partition templates
The invariants properties should be upheld in any partition template:

* It must contain at least one part
* It must not contain invalid Strftime formatters (`%3` is a good example I have used)
* A strftime formatter should be non-empty

I think the latter will involve attempting to format a timestamp with the input format spec, because the validation logic doesn't run until then.

### PR DESCRIPTION
**Pull request by carols10cents**
Fixes #7895.

On creation of custom partition templates, they are checked that they have:

- At least one part
- No more than 8 parts
- Only nonempty, valid strftime formats

### CODE REVIEW

**Review by domodwyer — APPROVED** [review]
Thank you for the fast turnaround!

File: data_types/src/partition_template.rs [hunk 1]
```diff
@@ -197,48 +238,118 @@ impl TablePartitionTemplateOverride {
...
-struct SerializationWrapper(Arc<proto::PartitionTemplate>);
-
-impl<DB> sqlx::Type<DB> for SerializationWrapper
-where
-    sqlx::types::Json<Self>: sqlx::Type<DB>,
-    DB: sqlx::Database,
-{
-    fn type_info() -> DB::TypeInfo {
-        <sqlx::types::Json<Self> as sqlx::Type<DB>>::type_info()
+mod serialization {
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply data_types/src/partition_template.rs:241]: I did indeed :)

File: data_types/src/partition_template.rs [hunk 2]
```diff
@@ -92,6 +92,39 @@ use generated_types::influxdata::iox::partition_template::v1 as proto;
 use once_cell::sync::Lazy;
 use percent_encoding::{percent_decode_str, AsciiSet, CONTROLS};
 use std::{borrow::Cow, sync::Arc};
+use thiserror::Error;
+
+/// TODO: Actually validate
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:97]: I just saw this - this can go right?! (I'm happy to remove it, just checking!)

Comment (carols10cents) [reply data_types/src/partition_template.rs:97]: WHOOPS yes!

**Pull request merged**